lib/kodama/client.rb

```diff
@@ -254,12 +278,31 @@ def callback(name, *args)
     # Set the next event position for the next iteration
     # Compare positions to avoid decreasing the position unintentionally
     # because next_position of Binlog::FormatEvent is always 0
-    def set_next_event_position(binlog_info, event)
-      if !event.kind_of?(Binlog::RotateEvent) && binlog_info.position.to_i < event.next_position.to_i
-        binlog_info.position = event.next_position
+    def set_next_event_position(binlog_info, cur_binlog_pos, event)
+      next_position = calc_next_position(cur_binlog_pos, event)
+      if !event.kind_of?(Binlog::RotateEvent) && binlog_info.position.to_i < next_position
+        binlog_info.position = next_position
+      end
+    end
+
+    def calc_next_position(cur_binlog_pos, event)
+      # Calculate next position from current position and event length if position is overflowed
+      if next_position_overflowed?(cur_binlog_pos, event)
+        # return 0 when next_position is 0
+        if event.next_position.to_i == 0
+          0
+        else
+          cur_binlog_pos + event.event_length
+        end
+      else
+        event.next_position
       end
     end
 
+    def next_position_overflowed?(cur_binlog_pos, event)
+      (event.next_position != 0 && event.next_position < cur_binlog_pos.to_i) || @current_position_overflowed
```

Comment:
Because 0 will be set to artificial events like Format, Rotate events, those are not existing in binlog. 

https://github.com/mysql/mysql-server/blob/5.7/sql/log_event.cc#L1208-L1212

```
    /*
      Artificial events are automatically generated and do not exist
      in master's binary log, so log_pos should be set to 0.
    */
```